pkg/lib/cockpit-components-terminal.tsx

```diff
@@ -103,13 +104,96 @@ const themes = {
  *
  * Also it is possible to set up theme by property 'theme'.
  */
-export class Terminal extends React.Component {
-    constructor(props) {
+
+export class TerminalState {
+    terminal: Term;
+    element: HTMLDivElement;
+    channel: cockpit.Channel<string>;
+
+    constructor(channel: cockpit.Channel<string>) {
+        this.terminal = new Term({
+            cols: 1,
+            rows: 1,
+            cursorBlink: true,
+            fontSize: 16,
+            fontFamily: 'Menlo, Monaco, Consolas, monospace',
+            screenReaderMode: true,
+        });
+        this.terminal.loadAddon(new CanvasAddon());
+        this.element = document.createElement("div");
+        this.channel = channel;
+    }
+
+    connectChannel() {
+        this.channel.addEventListener('message', (_event, data) => {
+            this.terminal.write(data);
+        });
+
+        this.terminal.onData(data => {
+            if (this.channel.valid) {
+                /* HACK: Ctrl+Space (and possibly other
+                 * characters) is a disaster: While it is U+00A0
+                 * in unicode, with an UTF-8 representation of
+                 * 0xC2A0, the "visible" string in JS is 0x00
+                 * (with TextEncoder, btoa(), and string
+                 * comparison). The internal representation
+                 * retains half of it, and trying to send it to
+                 * the websocket would result in a single 0xA0,
+                 * which is invalid UTF-8 (and causes the session
+                 * to crash). So intercept and ignore such broken
+                 * chars.  See
+                 * https://github.com/cockpit-project/cockpit/issues/21213 */
+                if (data === '\x00') {
+                    console.log("terminal: ignoring invalid input", data);
+                    return;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-2a204949-20250814-103842-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)